### PR TITLE
9785 add comment column in docker history command

### LINKED ISSUE #9785: Make comment message to be displayed in docker history command
I raised a discussion here

https://groups.google.com/forum/#!topic/docker-dev/AlbF1ij5oY0

<code>docker commit</code> command accept a <code>-m</code> parameter which allow users to add some messages to their image, just like which people would do while commit their code.

But anyway these messages are not read out when a user issue a <code>docker history</code> command to that image. 

By inspecting into source code, I found these messages were surely stored with images, so I made a patch to fetch them out. 

````patch
root@966e1e937644:/go/src/github.com/docker/docker# git diff 739d917..HEAD
diff --git a/api/client/commands.go b/api/client/commands.go
index a96089b..04d0496 100644
--- a/api/client/commands.go
+++ b/api/client/commands.go
@@ -1036,7 +1036,7 @@ func (cli *DockerCli) CmdHistory(args ...string) error {

 
        w := tabwriter.NewWriter(cli.out, 20, 1, 3, ' ', 0)
        if !*quiet {
-               fmt.Fprintln(w, "IMAGE\tCREATED\tCREATED BY\tSIZE")
+               fmt.Fprintln(w, "IMAGE\tCREATED\tCREATED BY\tSIZE\tCOMMENT")
        }
 
        for _, out := range outs.Data {
@@ -1055,7 +1055,8 @@ func (cli *DockerCli) CmdHistory(args ...string) error {

                        } else {
                                fmt.Fprintf(w, "%s\t", utils.Trunc(out.Get("CreatedBy"), 45))
                        }
-                       fmt.Fprintf(w, "%s\n", units.HumanSize(out.GetInt64("Size")))
+                       fmt.Fprintf(w, "%s\t", units.HumanSize(out.GetInt64("Size")))
+                        fmt.Fprintf(w, "%s\n", out.Get("Comment"))
                } else {
                        if *noTrunc {
                                fmt.Fprintln(w, outID)
diff --git a/graph/history.go b/graph/history.
go
index 2030c4c..853d6c6 100644
--- a/graph/history.go
+++ b/graph/history.go
@@ -36,6 +36,7 @@ func (s *TagStore) CmdHistory(job *engine.Job) engine.Status {

                out.Set("CreatedBy", strings.Join(img.ContainerConfig.Cmd, " "))
                out.SetList("Tags", lookupMap[img.ID])
                out.SetInt64("Size", img.Size)
+                out.Set("Comment", img.Comment)
                outs.Add(out)
                return nil
        })

root@966e1e937644:/go/src/github.com/docker/docker#
````

With this patch you can have a COMMENT column in the output of <code>docker history</code> command.

````shell
root@localhost docker]# ./docker history 2ac9d1098bf1
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
2ac9d1098bf1        12 weeks ago        /bin/bash                                       241.4 MB            Add Git and SVN
88d178232033        12 weeks ago        /bin/bash                                       186 MB              Add JDK1.7
4e5e740790d1        3 months ago        /bin/bash                                       535 MB              
88b42ffd1f7c        5 months ago        /bin/sh -c #(nop) ADD file:1fd8d7f9f6557cafc7   373.7 MB            
c69cab00d6ef        5 months ago        /bin/sh -c #(nop) MAINTAINER Lokesh Mandvekar   0 B                
511136ea3c5a        18 months ago                                                       0 B                 Imported from -


````

### PR DESCRIPTION
**Pull request by dove-young**
I added a new column COMMENT in output of <code>docker history</code> command to display messages would be submitted using <code>docker commit </code> command with <code>-m</code> parameter. This change will give an output like this:

```
./docker history 2ac9d1098bf1
IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
2ac9d1098bf1        3 months ago        /bin/bash                                       241.4 MB            Add Git and SVN
88d178232033        3 months ago        /bin/bash                                       186 MB              Add JDK1.7
```

Messages submitted in  <code>docker commit </code> command are very important to me, and to other users I think, to trace image changes in Docker just like code changes in Git. So it is a very good feature, but I am also surprised of those messages are not displayed in <code>docker history</code> command output. This problem makes these messages make no sense. 

This PR is used to fetch these messages and display them into <code>docker history</code> command. Thanks for your appreciate.

Closes #9785

### CODE REVIEW

**Comment by SvenDowideit** [comment]
can you please update the examples in `cli.md` and `docs/man/docker-history.1.md`

**Comment by dove-young** [comment]
@SvenDowideit thanks for you remainder.  the example mentioned in <code>#history</code> section in <code>cli.md</code> implicit <code>--no-trunc=true</code> by default. It looks very weird in normal screen size, and since <code>docker history</code> implicit a <code>--no-trunc=false</code> by default since version 1.1.2, I prefer to change those examples to <code>--no-trunc=true</code> format.

**Comment by SvenDowideit** [comment]
@dove-young yup - going with the defaults sounds like a plan :)

**Comment by SvenDowideit** [comment]
Docs LGTM - @fredlf @jamtur01

**Comment by jamtur01** [comment]
LGTM

File: integration-cli/docker_cli_history_test.go [hunk 1]
```diff
@@ -82,3 +83,52 @@ func TestHistoryNonExistentImage(t *testing.T) {
 	}
 	logDone("history - history on non-existent image must fail")
 }
+
+func TestHistoryImageWithComment(t *testing.T) {
+
+	// make a image through docker commit <container id> [ -m messages ]
+	runCmd := exec.Command(dockerBinary, "run", "-i", "-a", "stdin", "busybox", "echo", "foo")
```
Comment (icecrime) [inline integration-cli/docker_cli_history_test.go]: I'm assuming you specified `-a stdin` to avoid the echo messing with your command output. I don't think it's worth updating the PR, but it would be easier to have a trivial run such as:

    runCmd := exec.Command(dockerBinary, "run", "--name", "test", "busybox", "true")

And use the attributed name later on to refer to the container.

File: integration-cli/docker_cli_history_test.go [hunk 2]
```diff
@@ -82,3 +83,52 @@ func TestHistoryNonExistentImage(t *testing.T) {
...
+
+	commitCmd := exec.Command(dockerBinary, "commit", "-m=This is a comment", cleanedContainerID)
+	out, _, err = runCommandWithOutput(commitCmd)
+	if err != nil {
+		t.Fatalf("failed to commit container to image: %s, %v", out, err)
+	}
+
+	cleanedImageID := stripTrailingCharacters(out)
+	deleteContainer(cleanedContainerID)
+	logDone("history - commit a container with comment")
```
Comment (icecrime) [inline integration-cli/docker_cli_history_test.go]: That's a little early to print that the test passed :-) This should be moved at the very end of the function.

File: integration-cli/docker_cli_history_test.go [hunk 3]
```diff
@@ -82,3 +83,52 @@ func TestHistoryNonExistentImage(t *testing.T) {
...
+	historyCmd := exec.Command(dockerBinary, "history", cleanedImageID)
+	out, exitCode, err := runCommandWithOutput(historyCmd)
+	if err != nil || exitCode != 0 {
+		t.Fatalf("failed to get image history: %s, %v", out, err)
+	}
+
+	expectedValue := "This is a comment"
+
+	outputLine := strings.Split(out, "\n")[1]
+	outputTabs := regexp.MustCompile("  +").Split(outputLine, -1)
```
Comment (icecrime) [inline integration-cli/docker_cli_history_test.go]: This can be more easily achieved with [`strings.Fields`](http://golang.org/pkg/strings/#Fields).

File: integration-cli/docker_cli_history_test.go [hunk 4]
```diff
@@ -82,3 +83,52 @@ func TestHistoryNonExistentImage(t *testing.T) {
...
+	actualValue := outputTabs[len(outputTabs)-1]
+
+	if strings.Contains(actualValue, expectedValue) {
```
Comment (icecrime) [inline integration-cli/docker_cli_history_test.go]: Can we swap the condition? Like: `if !strings.Contains(...) { t.Fatalf() }` and then end with the `logDone` call?

File: integration-cli/docker_cli_history_test.go [hunk 5]
```diff
@@ -82,3 +83,52 @@ func TestHistoryNonExistentImage(t *testing.T) {
...
+		deleteImages(cleanedImageID)
```
Comment (icecrime) [inline integration-cli/docker_cli_history_test.go]: The call to `deleteImages()` should be deferred right after a successful commit. Here for example the image won't be cleaned up if the call to `history` happened to fail.

**Comment by icecrime** [comment]
These are my remarks:

1. Thanks a lot for your contribution! I like it :-)
2. I see some improvements on the test implementation
3. The PR contains 6 commits, which I'd rather see [squashed](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html) in 1

For other maintainers: yes this has UX implications, however it only adds a flag to what I consider to be a "second-class" command (namely `history`), so I'm +1 on the design.

**Comment by dove-young** [comment]
Thanks very much for your comments. I will follow them and update test. And BTW, would it break something if I rebase these commits into 1 on this branch?

**Comment by icecrime** [comment]
It shouldn't, and it's even recommended!

File: integration-cli/docker_cli_history_test.go (hunk 1 above)
Comment (dove-young) [inline integration-cli/docker_cli_history_test.go]: I tried this command line, but it failed the test. I'd like to keep existing command line.

        --- FAIL: TestHistoryImageWithComment (0.35 seconds)
                docker_cli_history_test.go:100: error thrown while waiting for container: , exit status 1

File: integration-cli/docker_cli_history_test.go (hunk 3 above)
Comment (dove-young) [inline integration-cli/docker_cli_history_test.go]: I prefer to use <code>regexp</code> to split each tabs since most tabs contain blank spaces. <code>strings.Fields</code> will slip in this condition and <code>strings.FieldsFunc</code> seems more complex than <code>regexp</code>

**Comment by fredlf** [comment]
@dove-young I think something went wrong with a rebase or some such. This PR currently wants to change 275 files!

**Comment by dove-young** [comment]
@fredif  You are right. There are only 5 files changed actually. You know, there were something happened. I tried to merge upstream changes in master HEAD into this branch continuously. And I used Squash on those commits merged in my branch in the rebase. I guess it is something wrong. 

````diff
diff --git a/api/client/commands.go b/api/client/commands.go
index 237caa6..8fd9564 100644
--- a/api/client/commands.go
+++ b/api/client/commands.go
@@ -1059,7 +1059,7 @@ func (cli *DockerCli) CmdHistory(args ...string) error {
 
 	w := tabwriter.NewWriter(cli.out, 20, 1, 3, ' ', 0)
 	if !*quiet {
-		fmt.Fprintln(w, "IMAGE\tCREATED\tCREATED BY\tSIZE")
+		fmt.Fprintln(w, "IMAGE\tCREATED\tCREATED BY\tSIZE\tCOMMENT")
 	}
 
 	for _, out := range outs.Data {
@@ -1078,7 +1078,8 @@ func (cli *DockerCli) CmdHistory(args ...string) error {
 			} else {
 				fmt.Fprintf(w, "%s\t", utils.Trunc(out.Get("CreatedBy"), 45))
 			}
-			fmt.Fprintf(w, "%s\n", units.HumanSize(float64(out.GetInt64("Size"))))
+			fmt.Fprintf(w, "%s\t", units.HumanSize(float64(out.GetInt64("Size"))))
+			fmt.Fprintf(w, "%s\n", out.Get("Comment"))
 		} else {
 			if *noTrunc {
 				fmt.Fprintln(w, outID)
diff --git a/docs/man/docker-history.1.md b/docs/man/docker-history.1.md
index 47350f8..545537c 100644
--- a/docs/man/docker-history.1.md
+++ b/docs/man/docker-history.1.md
@@ -26,11 +26,24 @@ Show the history of when and how an image was created.
    Only show numeric IDs. The default is *false*.
 
 # EXAMPLES
+
+## Show the history of images created through docker build command
+
     $ sudo docker history fedora
-    IMAGE          CREATED          CREATED BY                                      SIZE
+    IMAGE          CREATED          CREATED BY                                      SIZE                COMMENT
     105182bb5e8b   5 days ago       /bin/sh -c #(nop) ADD file:71356d2ad59aa3119d   372.7 MB
     73bd853d2ea5   13 days ago      /bin/sh -c #(nop) MAINTAINER Lokesh Mandvekar   0 B
-    511136ea3c5a   10 months ago                                                    0 B
+    511136ea3c5a   10 months ago                                                    0 B                 Imported from -
+
+## Show the history of images created through docker commit command
+`docker commit` command accepts a **-m** parameter to provide comment messages to the image. You can see these messages in image history.
+
+    $ sudo docker history docker:scm
+    IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
+    2ac9d1098bf1        3 months ago        /bin/bash                                       241.4 MB            Added Apache to Fedora base image
+    88b42ffd1f7c        5 months ago        /bin/sh -c #(nop) ADD file:1fd8d7f9f6557cafc7   373.7 MB            
+    c69cab00d6ef        5 months ago        /bin/sh -c #(nop) MAINTAINER Lokesh Mandvekar   0 B                 
+    511136ea3c5a        19 months ago                                                       0 B                 Imported from -
 
 # HISTORY
 April 2014, Originally compiled by William Henry (whenry at redhat dot com)
diff --git a/docs/sources/reference/commandline/cli.md b/docs/sources/reference/commandline/cli.md
index 47d32b4..ace1375 100644
--- a/docs/sources/reference/commandline/cli.md
+++ b/docs/sources/reference/commandline/cli.md
@@ -1021,13 +1021,23 @@ For example:
 To see how the `docker:latest` image was built:
 
     $ sudo docker history docker
-    IMAGE                                                              CREATED             CREATED BY                                                                                                                                                 SIZE
-    3e23a5875458790b7a806f95f7ec0d0b2a5c1659bfc899c89f939f6d5b8f7094   8 days ago          /bin/sh -c #(nop) ENV LC_ALL=C.UTF-8                                                                                                                       0 B
-    8578938dd17054dce7993d21de79e96a037400e8d28e15e7290fea4f65128a36   8 days ago          /bin/sh -c dpkg-reconfigure locales &&    locale-gen C.UTF-8 &&    /usr/sbin/update-locale LANG=C.UTF-8                                                    1.245 MB
-    be51b77efb42f67a5e96437b3e102f81e0a1399038f77bf28cea0ed23a65cf60   8 days ago          /bin/sh -c apt-get update && apt-get install -y    git    libxml2-dev    python    build-essential    make    gcc    python-dev    locales    python-pip   338.3 MB
-    4b137612be55ca69776c7f30c2d2dd0aa2e7d72059820abf3e25b629f887a084   6 weeks ago         /bin/sh -c #(nop) ADD jessie.tar.xz in /                                                                                                                   121 MB
-    750d58736b4b6cc0f9a9abe8f258cef269e3e9dceced1146503522be9f985ada   6 weeks ago         /bin/sh -c #(nop) MAINTAINER Tianon Gravi <admwiggin@gmail.com> - mkimage-debootstrap.sh -t jessie.tar.xz jessie http://http.debian.net/debian             0 B
-    511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158   9 months ago                                                                                                                                                                   0 B
+    IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
+    3e23a5875458        8 days ago          /bin/sh -c #(nop) ENV LC_ALL=C.UTF-8            0 B
+    8578938dd170        8 days ago          /bin/sh -c dpkg-reconfigure locales &&    loc   1.245 MB
+    be51b77efb42        8 days ago          /bin/sh -c apt-get update && apt-get install    338.3 MB
+    4b137612be55        6 weeks ago         /bin/sh -c #(nop) ADD jessie.tar.xz in /        121 MB
+    750d58736b4b        6 weeks ago         /bin/sh -c #(nop) MAINTAINER Tianon Gravi <ad   0 B
+    511136ea3c5a        9 months ago                                                        0 B                 Imported from -
+
+To see how the `docker:apache` image which was commited from a container:
+
+    $ sudo docker history docker:scm
+    IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
+    2ac9d1098bf1        3 months ago        /bin/bash                                       241.4 MB            Added Apache to Fedora base image
+    88b42ffd1f7c        5 months ago        /bin/sh -c #(nop) ADD file:1fd8d7f9f6557cafc7   373.7 MB            
+    c69cab00d6ef        5 months ago        /bin/sh -c #(nop) MAINTAINER Lokesh Mandvekar   0 B                 
+    511136ea3c5a        19 months ago                                                       0 B                 Imported from -
+
 
 ## images
 
diff --git a/graph/history.go b/graph/history.go
index 3563406..a069c39 100644
--- a/graph/history.go
+++ b/graph/history.go
@@ -36,6 +36,7 @@ func (s *TagStore) CmdHistory(job *engine.Job) engine.Status {
 		out.Set("CreatedBy", strings.Join(img.ContainerConfig.Cmd, " "))
 		out.SetList("Tags", lookupMap[img.ID])
 		out.SetInt64("Size", img.Size)
+		out.Set("Comment", img.Comment)
 		outs.Add(out)
 		return nil
 	})
diff --git a/integration-cli/docker_cli_history_test.go b/integration-cli/docker_cli_history_test.go
index 3ae9ffb..a6dcdfd 100644
--- a/integration-cli/docker_cli_history_test.go
+++ b/integration-cli/docker_cli_history_test.go
@@ -3,6 +3,7 @@ package main
 import (
 	"fmt"
 	"os/exec"
+	"regexp"
 	"strings"
 	"testing"
 )
@@ -82,3 +83,49 @@ func TestHistoryNonExistentImage(t *testing.T) {
 	}
 	logDone("history - history on non-existent image must fail")
 }
+
+func TestHistoryImageWithComment(t *testing.T) {
+
+	// make a image through docker commit <container id> [ -m messages ]
+	runCmd := exec.Command(dockerBinary, "run", "-i", "-a", "stdin", "busybox", "echo", "foo")
+	out, _, _, err := runCommandWithStdoutStderr(runCmd)
+	if err != nil {
+		t.Fatalf("failed to run container: %s, %v", out, err)
+	}
+
+	cleanedContainerID := stripTrailingCharacters(out)
+
+	waitCmd := exec.Command(dockerBinary, "wait", cleanedContainerID)
+	if _, _, err = runCommandWithOutput(waitCmd); err != nil {
+		t.Fatalf("error thrown while waiting for container: %s, %v", out, err)
+	}
+
+	commitCmd := exec.Command(dockerBinary, "commit", "-m=This is a comment", cleanedContainerID)
+	out, _, err = runCommandWithOutput(commitCmd)
+	if err != nil {
+		t.Fatalf("failed to commit container to image: %s, %v", out, err)
+	}
+
+	cleanedImageID := stripTrailingCharacters(out)
+	deleteContainer(cleanedContainerID)
+	defer deleteImages(cleanedImageID)
+
+	// test docker history <image id> to check comment messages
+	historyCmd := exec.Command(dockerBinary, "history", cleanedImageID)
+	out, exitCode, err := runCommandWithOutput(historyCmd)
+	if err != nil || exitCode != 0 {
+		t.Fatalf("failed to get image history: %s, %v", out, err)
+	}
+
+	expectedValue := "This is a comment"
+
+	outputLine := strings.Split(out, "\n")[1]
+	outputTabs := regexp.MustCompile("  +").Split(outputLine, -1)
+	actualValue := outputTabs[len(outputTabs)-1]
+
+	if !strings.Contains(actualValue, expectedValue) {
+		t.Fatalf("Expected comments \"%s\", but found \"%s\"", expectedValue, actualValue)
+	}
+
+	logDone("history - history on image with comment")
+}
````diff

**Comment by dove-young** [comment]
Changed back to pure my changes.

**Comment by jfrazelle** [comment]
ping @stevvooe does the new image format have a notion of a description/message like this?

**Comment by stevvooe** [comment]
@jfrazelle Yes, it does (docker/distribution#62). However, history would have to be implemented with metadata.

@jlhawn Thoughts?

**Comment by jlhawn** [comment]
@jfrazelle @stevvooe Docker Image JSONs don't exist in the new manifest format. It's up to the Docker project to decide what data to put in the Application Data field: https://github.com/jlhawn/distribution/blob/generic_manifest/doc/spec/manifest.md#example-manifest

If commit comments are deemed important enough, you could put them in the application data, otherwise I recommend they go in the Meta data field. In the example manifest I linked above, you can see the `buildSteps` Meta data field which is a JSON array of the Dockerfile commands. This is just an example, but that is pretty much what `docker history ` prints out today in the `CREATED BY` column, i.e., what command was run to produce that image layer.

While I'm at it, I want to draw your attention to this Issue I opened a while ago: https://github.com/docker/docker/issues/10753

The docker project NEEDS to decide on a media type for the `application` field in the upcoming manifest format (linked above). The one I made up in my example (`application/vnd.docker.container.image.params.v1+json`) is based on the fields used by our current image format. If you want to discuss this, please do so on #10753 :+1:

**Comment by stevvooe** [comment]
@jlhawn In your opinion, should adding this comment field be part of the effort in #9882? This seems like a slippery slope (why not add `<whatever field I care about the most>`?).

**Comment by jlhawn** [comment]
@stevvooe I don't think it should affect #9882 at all. Each "layer" comment is meant to be distinct and a Dockerfile `COMMENT` command would create it's own "layer".

The `docker history` command itself is weird going into the future because it's focused around the 'layer history' of an image and you need all of the layers to know everything. I think what people really use it for is "Build History", i.e., what were the steps that went into building the image? This is something which can be collected in one place in the image manifest's Meta data field.


I think @dove-young submitted this to solve the problem that for users who build images using `docker commit` the `CREATED BY` column isn't very helpful, since it's just whatever the CMD was when the container was created. I could do:

```
docker run -it --name my_build_container ubuntu:utopic bash
```

then do a whole bunch of things in the container like manually install dependencies and my application, then

```
docker commit -m "I did all these things" my_build_container jlhawn/my-app:latest
```

but the `docker history jlhawn/my-app:latest` output would look like this:

```
IMAGE               CREATED             CREATED BY                                      SIZE
822a18b038e8        6 seconds ago       bash                                            0 B
cfaba6b5fefe        6 weeks ago         /bin/sh -c #(nop) CMD [/bin/bash]               0 B
ea198eaf8e7b        6 weeks ago         /bin/sh -c sed -i 's/^#\s*\(deb.*universe\)$/   1.895 kB
36accde0a93b        6 weeks ago         /bin/sh -c echo '#!/bin/sh' > /usr/sbin/polic   215 kB
ef3ba9f35b97        6 weeks ago         /bin/sh -c #(nop) ADD file:eadcbf32d261603b43   194.2 MB
511136ea3c5a        21 months ago                                                       0 B
```

Which is misleading... That layer wasn't created by `bash`, it was created by "all these things [I did]".

TLDR: I like this PR and adding a `COMMENT` column is a good idea. No, there probably should *not* be a `COMMENT` Dockerfile command.

**Comment by tiborvass** [comment]
I think the design makes sense and agree with @jlhawn.

I'm pushing this to code review as I doubt anyone would object, but feel free to contradict me!

**Comment by tiborvass** [comment]
@dove-young can you make sure to update the API along with the API docs? Thanks!

File: docs/sources/reference/commandline/cli.md [hunk 6]
```diff
@@ -945,13 +945,23 @@ For example:
...
-    511136ea3c5a64f264b78b5433614aec563103b4d4702f3ba7d4d2698e22c158   9 months ago                                                                                                                                                                   0 B
+    IMAGE               CREATED             CREATED BY                                      SIZE                COMMENT
+    3e23a5875458        8 days ago          /bin/sh -c #(nop) ENV LC_ALL=C.UTF-8            0 B
+    8578938dd170        8 days ago          /bin/sh -c dpkg-reconfigure locales &&    loc   1.245 MB
+    be51b77efb42        8 days ago          /bin/sh -c apt-get update && apt-get install    338.3 MB
+    4b137612be55        6 weeks ago         /bin/sh -c #(nop) ADD jessie.tar.xz in /        121 MB
+    750d58736b4b        6 weeks ago         /bin/sh -c #(nop) MAINTAINER Tianon Gravi <ad   0 B
+    511136ea3c5a        9 months ago                                                        0 B                 Imported from -
+
+To see how the `docker:apache` image which was commited from a container:
```
Comment (fredlf) [inline docs/sources/reference/commandline/cli.md]: Need to rephrase this so it's grammatically correct. Suggest something like:

To see how the `docker:apache` image was added to a container's base image:

File: docs/man/docker-history.1.md [hunk 7]
```diff
@@ -22,11 +22,24 @@ Show the history of when and how an image was created.
...
     $ sudo docker history fedora
-    IMAGE          CREATED          CREATED BY                                      SIZE
+    IMAGE          CREATED          CREATED BY                                      SIZE                COMMENT
     105182bb5e8b   5 days ago       /bin/sh -c #(nop) ADD file:71356d2ad59aa3119d   372.7 MB
     73bd853d2ea5   13 days ago      /bin/sh -c #(nop) MAINTAINER Lokesh Mandvekar   0 B
-    511136ea3c5a   10 months ago                                                    0 B
+    511136ea3c5a   10 months ago                                                    0 B                 Imported from -
+
+## Show the history of images created through docker commit command
+`docker commit` command accepts a **-m** parameter to provide comment messages to the image. You can see these messages in image history.
```
Comment (fredlf) [inline docs/man/docker-history.1.md]: Let's simplify and clarify this:

The `docker commit` command has a **-m** flag for adding comments to the image. These comments will be displayed in the image history.

**Comment by icecrime** [comment]
Why are we reviewing docs?

**Comment by dove-young** [comment]
Thanks for your comments and sorry for late response. I've updated documents by the comments from @fredlf 

And I have no idea why it failed in building.  I've tried to merge this change with master HEAD at Jan 20 and it caused 275 files been changed.  Should I just keep it as-is?

**Comment by LK4D4** [comment]
So, if all like it, let's do it.
ping @tiborvass @crosbymichael for removing UX label

**Comment by tiborvass** [comment]
@LK4D4 I already removed design review label and this PR is in code review.
@dove-young let us know if you're tired of rebasing (would need another one) :P, we can carry the PR.

**Comment by thaJeztah** [comment]
ping @dove-young (https://github.com/docker/docker/pull/9916#issuecomment-88238441), can you let us know if you want to rebase, or want @tiborvass to carry it? Thanks!

**Comment by tiborvass** [comment]
Carried in #12210
Thanks @thaJeztah
